In [ ]:
import pandas as pd

# Читаем файл melb_data.csv
melb_data = pd.read_csv('data/melb_data.csv', sep=',')

Вывод первых и последних строк

Для вывода первых и последних строк в DataFrame есть методы head() и tail(), которые возвращают n первых и n последних строк таблицы соответственно (по умолчанию n = 5)

In [2]:
# Вывод первых 5 строк таблицы (индексы с 0 по 4)
display(melb_data.head())

,index,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,...,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount,Coordinates
0,0,Abbotsford,85 Turner St,2,h,1480000.0,S,Biggin,3/12/2016,2.5,...,1.0,202.0,126.0,1970.0,Yarra,-37.7996,144.9984,Northern Metropolitan,4019.0,"-37.7996, 144.9984"
1,1,Abbotsford,25 Bloomburg St,2,h,1035000.0,S,Biggin,4/02/2016,2.5,...,0.0,156.0,79.0,1900.0,Yarra,-37.8079,144.9934,Northern Metropolitan,4019.0,"-37.8079, 144.9934"
2,2,Abbotsford,5 Charles St,3,h,1465000.0,SP,Biggin,4/03/2017,2.5,...,0.0,134.0,150.0,1900.0,Yarra,-37.8093,144.9944,Northern Metropolitan,4019.0,"-37.8093, 144.9944"
3,3,Abbotsford,40 Federation La,3,h,850000.0,PI,Biggin,4/03/2017,2.5,...,1.0,94.0,126.0,1970.0,Yarra,-37.7969,144.9969,Northern Metropolitan,4019.0,"-37.7969, 144.9969"
4,4,Abbotsford,55a Park St,4,h,1600000.0,VB,Nelson,4/06/2016,2.5,...,2.0,120.0,142.0,2014.0,Yarra,-37.8072,144.9941,Northern Metropolitan,4019.0,"-37.8072, 144.9941"


In [3]:
# Вывод семи последних строк таблицы
display(melb_data.tail(7))

,index,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,...,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount,Coordinates
13573,13573,Werribee,5 Nuragi Ct,4,h,635000.0,S,hockingstuart,26/08/2017,14.7,...,1.0,662.0,172.0,1980.0,NaN,-37.89327,144.64789,Western Metropolitan,16166.0,"-37.89327, 144.64789"
13574,13574,Westmeadows,9 Black St,3,h,582000.0,S,Red,26/08/2017,16.5,...,2.0,256.0,126.0,1970.0,NaN,-37.67917,144.89390,Northern Metropolitan,2474.0,"-37.67917, 144.8939"
13575,13575,Wheelers Hill,12 Strada Cr,4,h,1245000.0,S,Barry,26/08/2017,16.7,...,2.0,652.0,126.0,1981.0,NaN,-37.90562,145.16761,South-Eastern Metropolitan,7392.0,"-37.90562, 145.16761"
13576,13576,Williamstown,77 Merrett Dr,3,h,1031000.0,SP,Williams,26/08/2017,6.8,...,2.0,333.0,133.0,1995.0,NaN,-37.85927,144.87904,Western Metropolitan,6380.0,"-37.85927, 144.87904"
13577,13577,Williamstown,83 Power St,3,h,1170000.0,S,Raine,26/08/2017,6.8,...,4.0,436.0,126.0,1997.0,NaN,-37.85274,144.88738,Western Metropolitan,6380.0,"-37.85274, 144.88738"
13578,13578,Williamstown,96 Verdon St,4,h,2500000.0,PI,Sweeney,26/08/2017,6.8,...,5.0,866.0,157.0,1920.0,NaN,-37.85908,144.89299,Western Metropolitan,6380.0,"-37.85908, 144.89299"
13579,13579,Yarraville,6 Agnes St,4,h,1285000.0,SP,Village,26/08/2017,6.3,...,1.0,362.0,112.0,1920.0,NaN,-37.81188,144.88449,Western Metropolitan,6543.0,"-37.81188, 144.88449"


Размерность таблицы

Размерность таблицы (количество строк и столбцов) можно узнать с помощью атрибута shape, который возвращает кортеж с количеством строк и столбцов.

In [4]:
display(melb_data.shape)

(13580, 23)

Признаки — столбцы в таблице DataFrame.

Получение информации о столбцах

Чтобы получить более детальную информацию о признаках (столбцах) таблицы, можно использовать метод DataFrame info().

Метод info() выводит:

— информацию об индексах;
— информацию об общем количестве столбцов;
— таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце и типе данных столбца (Dtype), количестве столбцов, в которых используется определенный тип данных;
— количество оперативной памяти в мегабайтах, которое тратится на хранение данных.

In [5]:
display(melb_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

None

Из вывода метода info() становится понятно, что в нашей таблице есть столбец CouncilArea с пропущенными значениями — количество непустых значений меньше чем количество строк в таблице.

Пустыми, или пропущенными, значениями называются значения в ячейках таблицы, которые не заполнены по какой-либо причине, то есть на их месте стоит пустое место. В Pandas такие значения обозначаются символом NaN (Not-a-Number)

Некоторые признаки кодируются с некорректным типом данных. Например, признаки Car, Bedroom, Bathroom, Propertycount обозначают количество (целое), но кодируются с плавающей точкой (float64). 
Год постройки (YearBuilt) тоже представлен в формате с плавающей точкой.

Чтобы выполнить преобразование типа данных столбца, можно воспользоваться методом astype().

In [6]:
melb_data['Car'] = melb_data['Car'].astype('int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('int64')
display(melb_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

None

Основные статистические свойства столбцов таблицы можно вывести с помощью метода describe().

По умолчанию метод работает с числовыми (int64, float64) столбцами, и показывает число непустых значений (count), среднее (mean), стандартное отклонение (std), минимальное значение (min), квантили уровней 0.25, 0.5 (медиана) и 0.75 (25%, 50%, 75%), и максимальное значение (max) для каждого столбца исходной таблицы.

In [7]:
# Вывод значений статистических параметров для столбцов Distance 
# (расстояние от недвижимости до центра Мельбурна), BuildingArea (площадь здания) и Price (цена)
melb_data.describe().loc[:, ['Distance', 'BuildingArea', 'Price']]

,Distance,BuildingArea,Price
count,13580.000000,13580.000000,1.358000e+04
mean,10.137776,139.633972,1.075684e+06
std,5.868725,392.217403,6.393107e+05
min,0.000000,0.000000,8.500000e+04
25%,6.100000,122.000000,6.500000e+05
50%,9.200000,126.000000,9.030000e+05
75%,13.000000,129.940000,1.330000e+06
max,48.100000,44515.000000,9.000000e+06


Выводы:

— в данном наборе данных представлены объекты, удаленные от центра города на расстояние 48 км (столбец Distance);
— в таблице представлены участки, на которых отсутствует само здание (BuildingArea = 0), то есть владельцы продавали участок земли без строений на нем;
— большая часть объектов (75%) имеет площадь здания (BuildingArea) до 130 квадратных метров, но в таблице есть здание с площадью в 44 515 квадратных метров;
— диапазон цен на недвижимость (Price) варьируется от 85 тысяч до 9 миллионов австралийских долларов.

Метод describe() можно применять не только к числовым признакам. С помощью параметра include можно указать тип данных, для которого нужно вывести описательную информацию.

Например, для типа данных object метод describe() возвращает DataFrame, в котором указаны:

- количество непустых строк (count);
- количество уникальных значений (unique);
- самое частое значение — мода (top);
- частота — объем использования — этого значения (freq) для каждого столбца типа object исходной таблицы.

In [8]:
melb_data.describe(include=['object'])

,Suburb,Address,Type,Method,SellerG,Date,CouncilArea,Regionname,Coordinates
count,13580,13580,13580,13580,13580,13580,12211,13580,13580
unique,314,13378,3,5,268,58,33,8,13097
top,Reservoir,36 Aberfeldie St,h,S,Nelson,27/05/2017,Moreland,Southern Metropolitan,"-37.8361, 144.9966"
freq,359,3,9449,9022,1565,473,1163,4695,12


- наибольшее количество проданных объектов (359) находилось в пригороде Reservoir (столбец Suburb);
- самый популярный тип дома — h (дом, вилла, коттедж) — столбец Type (тип здания);
- в наших данных нам известно о 268 различных риэлторских компаниях, однако самой главной в этом бизнесе является компания Nelson — они продали 1565 различных домов (столбец SellerG);
- в нашей таблице содержится информация за 58 дней, при этом наибольшее число продаж (473) пришлось на 27 мая 2017 года (столбец Date);

Получение частоты уникальных значений в столбце

Чтобы определить, сколько раз в столбце повторяется каждый из вариантов значений (то есть определить частоту для каждого уникального значения), используется метод value_const().

Данный метод возвращает объект Series, в котором в качестве индексов выступают уникальные категории столбца, а значениями — соответствующая им частота.

In [9]:
# Пример частоты названий районов
melb_data['Regionname'].value_counts()

Regionname
Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: count, dtype: int64

Чтобы сделать вывод более интерпретируемым и понятным, можно использовать параметр normalize. При установке значения этого параметра на True результат будет представляться в виде доли (относительной частоты).

In [10]:
melb_data['Regionname'].value_counts(normalize=True)

Regionname
Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: proportion, dtype: float64

Из результатов следует вывод — наименьшее количество объектов (менее 1%) было продано в районах Victoria, а наибольшее — в районах Metropolitan.

Object — тип данных, которым в Pandas кодируются текстовые или другие нечисловые значения (кортежи, списки и тд)

In [13]:
melb_data.describe()

,index,Rooms,Price,Distance,Postcode,Bedroom,Bathroom,Car,Landsize,BuildingArea,YearBuilt,Lattitude,Longtitude,Propertycount
count,13580.000000,13580.000000,1.358000e+04,13580.000000,13580.000000,13580.000000,13580.000000,13580.000000,13580.000000,13580.000000,13580.000000,13580.000000,13580.000000,13580.000000
mean,6789.500000,2.937997,1.075684e+06,10.137776,3105.301915,2.914728,1.534242,1.611856,558.416127,139.633972,1966.788218,-37.809203,144.995216,7454.417378
std,3920.352663,0.955748,6.393107e+05,5.868725,90.676964,0.965921,0.691712,0.960793,3990.669241,392.217403,29.088642,0.079260,0.103916,4378.581772
min,0.000000,1.000000,8.500000e+04,0.000000,3000.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1196.000000,-38.182550,144.431810,249.000000
25%,3394.750000,2.000000,6.500000e+05,6.100000,3044.000000,2.000000,1.000000,1.000000,177.000000,122.000000,1960.000000,-37.856822,144.929600,4380.000000
50%,6789.500000,3.000000,9.030000e+05,9.200000,3084.000000,3.000000,1.000000,2.000000,440.000000,126.000000,1970.000000,-37.802355,145.000100,6555.000000
75%,10184.250000,3.000000,1.330000e+06,13.000000,3148.000000,3.000000,2.000000,2.000000,651.000000,129.940000,1975.000000,-37.756400,145.058305,10331.000000
max,13579.000000,10.000000,9.000000e+06,48.100000,3977.000000,20.000000,8.000000,10.000000,433014.000000,44515.000000,2018.000000,-37.408530,145.526350,21650.000000


Дополнительные выводы:

- половина проданных зданий находится на расстоянии от 6.1 до 13 километров от центра города (Distance);
- в данных есть как минимум одно жилое помещение, которое находится ровно в той точке, которая считается центром Мельбурна (строка min, столбец Distance).

In [14]:
melb_data.describe(include=['object'])

,Suburb,Address,Type,Method,SellerG,Date,CouncilArea,Regionname,Coordinates
count,13580,13580,13580,13580,13580,13580,12211,13580,13580
unique,314,13378,3,5,268,58,33,8,13097
top,Reservoir,36 Aberfeldie St,h,S,Nelson,27/05/2017,Moreland,Southern Metropolitan,"-37.8361, 144.9966"
freq,359,3,9449,9022,1565,473,1163,4695,12


Выводы:

- Мельбурн и его пригороды разделены как минимум на 8 районов (строка unique, столбец Regionname);
- у некоторых различных адресов в таблице совпадают географические координаты (строка unique в столбце Adress содержит больше значений, чем в столбце Coordinates). Это значит, что некоторые различные адреса помечены одинаковыми географическими координатами;
- в пригороде Reservoir расположено наибольшее (строка top, столбец Suburb) число проданных объектов - 359 (строка freq, столбец Suburb).

In [27]:
display(round(melb_data['Type'].value_counts(normalize=True) * 100).astype('Int16'))

Type
h    70
u    22
t     8
Name: proportion, dtype: Int16